scrapy/resolver.py

```diff
@@ -1,14 +1,86 @@
-from twisted.internet import defer
+import socket
+from twisted.python import failure
+from twisted.internet import defer, threads
+from twisted.internet.error import DNSLookupError
 from twisted.internet.base import ThreadedResolver
 
 from scrapy.utils.datatypes import LocalCache
 
+
+IPV4_ONLY = 4
+IPV6_ONLY = 6
+IPV4_PREFERRED = 46
+IPV6_PREFERRED = 64
+
+
+class IPv6ThreadedResolver(ThreadedResolver):
+    """ThreadedResolver calling socket.getaddrinfo for IPv6 capability.
+
+    Uses a threadpool to perform name lookups without blocking the
+    reactor thread. The underlying socket.getaddrinfo is a blocking call.
+    """
+
+    def __init__(self, reactor, protocol_preference=IPV4_PREFERRED):
+        super(IPv6ThreadedResolver, self).__init__(reactor)
+        # settings.getint() defaults to 0 for unset values
```

Comment:
Meaning we can depend on `DNS_PROTOCOL_PREFERENCE` being set and `settings.getint('DNS_PROTOCOL_PREFERENCE')` never returning `0`?
If you think so, I'll remove the check.